source/common/http/match_delegate/config.h

```diff
@@ -104,17 +104,54 @@ class DelegatingStreamFilter : public Logger::Loggable<Logger::Id::http>,
 };
 
 class MatchDelegateConfig
-    : public Extensions::HttpFilters::Common::ExceptionFreeFactoryBase<
+    : public Extensions::HttpFilters::Common::CommonFactoryBase<
           envoy::extensions::common::matching::v3::ExtensionWithMatcher,
-          envoy::extensions::common::matching::v3::ExtensionWithMatcherPerRoute> {
+          envoy::extensions::common::matching::v3::ExtensionWithMatcherPerRoute>,
+      public Server::Configuration::NamedHttpFilterConfigFactory,
+      public Server::Configuration::UpstreamHttpFilterConfigFactory {
 public:
   // TODO(wbpcode): move this filter to 'source/extensions/filters/http'.
-  MatchDelegateConfig() : ExceptionFreeFactoryBase("envoy.filters.http.match_delegate") {}
+  MatchDelegateConfig()
+      : CommonFactoryBase<envoy::extensions::common::matching::v3::ExtensionWithMatcher,
+                          envoy::extensions::common::matching::v3::ExtensionWithMatcherPerRoute>(
+            "envoy.filters.http.match_delegate") {}
+
+  absl::StatusOr<Envoy::Http::FilterFactoryCb>
+  createFilterFactoryFromProto(const Protobuf::Message& proto_config,
+                               const std::string& stats_prefix,
+                               Server::Configuration::FactoryContext& context) override {
+    return createFilterFactoryFromProtoTyped(
+        MessageUtil::downcastAndValidate<
+            const envoy::extensions::common::matching::v3::ExtensionWithMatcher&>(
+            proto_config, context.messageValidationVisitor()),
+        stats_prefix, context);
+  }
+
+  absl::StatusOr<Envoy::Http::FilterFactoryCb>
+  createFilterFactoryFromProto(const Protobuf::Message& proto_config,
+                               const std::string& stats_prefix,
+                               Server::Configuration::UpstreamFactoryContext& context) override {
+    return createFilterFactoryFromProtoTyped(
+        MessageUtil::downcastAndValidate<
+            const envoy::extensions::common::matching::v3::ExtensionWithMatcher&&>(
+            proto_config, context.serverFactoryContext().messageValidationVisitor()),
+        stats_prefix, context);
+  }
 
 private:
   absl::StatusOr<Envoy::Http::FilterFactoryCb> createFilterFactoryFromProtoTyped(
       const envoy::extensions::common::matching::v3::ExtensionWithMatcher& proto_config,
-      const std::string&, Server::Configuration::FactoryContext& context) override;
+      const std::string& prefix, Server::Configuration::FactoryContext& context);
+  absl::StatusOr<Envoy::Http::FilterFactoryCb> createFilterFactoryFromProtoTyped(
```

Comment:
I would rather let the compiler capture most of the problems at compile time instead of run time. Thus I would like to avoid such kind of std::is_same() check.

Reply:
std::is_same is a compile-time type trait

Reply:
I kind of prefer polymorphism in such case.  I guess either should work.